tests/kernel/workq/work_queue_api/src/main.c

```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+ *   sync_sema with initial count 0.
+ * - Handler function gives semaphore, then we wait for that semaphore
+ *   from the test function body.
+ * - If semaphore was obtained successfully, test passed.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_work_item_supplied_with_func(void)
+{
+	u32_t sem_count = 0;
+
+	k_sem_reset(&sync_sema);
+
+	/**TESTPOINT: init work item with a user-defined function*/
+	k_work_init(&work_item, common_work_handler);
+	k_work_submit_to_queue(&workq, &work_item);
+
+	k_sem_take(&sync_sema, K_FOREVER);
+	sem_count = k_sem_count_get(&sync_sema);
+	zassert_equal(sem_count, COM_SEM_INIT_VAL, NULL);
+}
+
+/* Two handler functions fifo_work_first() and fifo_work_second
+ * are made for two work items to test first in, first out.
+ * It tests workqueue thread process work items
+ * in first in, first out manner
+ */
+void fifo_work_first(struct k_work *unused)
+{
+	k_sem_take(&sema_fifo_one, K_FOREVER);
+	k_sem_give(&sema_fifo_two);
+}
+
+void fifo_work_second(struct k_work *unused)
+{
+	k_sem_take(&sema_fifo_two, K_FOREVER);
+}
+
+/**
+ * @brief Test kernel process work items in fifo way
+ * @details To test it we use 2 functions-handlers.
+ * - fifo_work_first() added to the work_item_1 and fifo_work_second()
+ * added to the work_item_2.
+ * - We expect that firstly should run work_item_1, and only then
+ * will run work_item_2.
+ * - To test it, we initialize semaphore sema_fifo_one
+ * with count 1(available) and fifo_work_first() takes that semaphore.
+ * fifo_work_second() is waiting for the semaphore sema_fifo_two,
+ * which will be given from function fifo_work_first().
+ * - If work_item_2() will try to be executed before work_item_1(),
+ * will happen a timeout error.
+ * - Because sema_fifo_two will be never obtained by fifo_work_second()
+ * due to K_FOREVER macro in it while waiting for the semaphore.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_process_work_items_fifo(void)
+{
+	k_work_init(&work_item_1, fifo_work_first);
+	k_work_init(&work_item_2, fifo_work_second);
+
+	/**TESTPOINT: submit work items to the queue in fifo manner*/
+	k_work_submit_to_queue(&workq, &work_item_1);
+	k_work_submit_to_queue(&workq, &work_item_2);
+}
+
+/**
+ * @brief Test kernel support scheduling work item that is to be processed
+ * after user defined period of time
+ * @details
+ * - For that test is using semaphore sync_sema, with initial count 0.
+ * - In that test we measure actual spent time and compare it with time
+ * which was measured by function k_delayed_work_remaining_get().
+ * - Using system clocks we measure actual spent time
+ * in the period between delayed work submitted and delayed work
+ * executed.
+ * - To know that delayed work was executed, we use semaphore.
+ * - Right after semaphore was given from handler function, we stop
+ * measuring actual time.
+ * - Then compare results.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_sched_delayed_work_item(void)
```

Comment:
I didn't find the test which checks that entered by user time for delay is correctly delayed in real program.

Reply:
I checked code and found test `test_delayed_work_submit_to_queue_thread` that is using function `tdelayed_work_submit_1` which tests the delayed time. But that function uses time got from a `k_delayed_work_remaining_get(w); ` as a default correct constant, and didn't try to verify that delayed time return from that function is correct and and least is in the limits of the real measurement. Also that test has some magic code line 
`timeout_ticks -= z_ms_to_ticks(15);` What means 15?
And I don't understand how that test verifies if the delayed time given by `k_delayed_work_remaining_get()` is really that time which was delayed by application in real life
 My test verifies that `k_delayed_work_remaining_get()` returned the correct number and that number is in the limits of the measured time using `k_cycle_get_32(); ` Unfortunately my test can't verify 100% that user defined delay of 100ms was in real life 100ms, because different arch use different timers, but it can definitely verify that delay in real life was 100ms +- delta 10ms.

I can keep old test, but I think necessary to modify it too. At least remove that magic line of code `timeout_ticks -= z_ms_to_ticks(15);` or make constant like I did with my delta constant.

Reply:
Heh, this is fine.  I didn't mean to send you crawling over the test suite (and, yes, a lot of our test code is of questionable quality).  There's no problem with adding a small and simple case for an obvious condition that isn't already tested in an obvious way.